compiler/p/codegen/OMRTreeEvaluator.cpp

```diff
@@ -3512,6 +3512,128 @@ TR::Register *OMR::Power::TreeEvaluator::vnegDoubleHelper(TR::Node *node, TR::Co
+       return TR::TreeEvaluator::inlineVectorBinaryOp(node, cg, TR::InstOpCode::vminsd);
+     case TR::Float:
+       return TR::TreeEvaluator::inlineVectorBinaryOp(node, cg, TR::InstOpCode::vminfp);
+     case TR::Double:
+       return TR::TreeEvaluator::vminDoubleHelper(node, cg);
+     default:
+       TR_ASSERT(false, "unrecognized vector type %s\n", node->getDataType().toString()); return NULL;
+     }
+   }
+
+TR::Register *OMR::Power::TreeEvaluator::vminDoubleHelper(TR::Node *node, TR::CodeGenerator *cg)
+   {
+   TR::Node *firstChild = node->getFirstChild();
+   TR::Node *secondChild = node->getSecondChild();
+
+   TR::Register *firstReg = cg->evaluate(firstChild);
+   TR::Register *secondReg = cg->evaluate(secondChild);
+
+   TR::Register *resReg = cg->allocateRegister(TR_VRF);
+
+   TR::Register *cmpReg = cg->allocateRegister(TR_VRF);
+   TR::Register *tempA = cg->allocateRegister(TR_VRF);
+   TR::Register *tempB = cg->allocateRegister(TR_VRF);
+
+   node->setRegister(resReg);
+
+   //check if first operand is NaN (all 0's if yes, all 1's if no), and if so, convert second operand to NaN
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::xvcmpeqdp, node, cmpReg, firstReg, firstReg);
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::xxlorc, node, tempA, secondReg, cmpReg);
+
+   //check if second operand is NaN (all 0's if yes, all 1's if no), and if so, convert first operand to NaN
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::xvcmpeqdp, node, cmpReg, secondReg, secondReg);
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::xxlorc, node, tempB, firstReg, cmpReg);
+
+   //VSX minimum (will return NaN if both operands are NaN)
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::xvmindp, node, resReg, tempA, tempB);
+
+   cg->stopUsingRegister(cmpReg);
+   cg->stopUsingRegister(tempA);
+   cg->stopUsingRegister(tempB);
+
+   cg->decReferenceCount(firstChild);
+   cg->decReferenceCount(secondChild);
+
+   return resReg;
+   }
+
+TR::Register *OMR::Power::TreeEvaluator::vmaxEvaluator(TR::Node *node, TR::CodeGenerator *cg)
+   {
+   TR_ASSERT_FATAL_WITH_NODE(node, node->getDataType().getVectorLength() == TR::VectorLength128,
+                   "Only 128-bit vectors are supported %s", node->getDataType().toString());
+
+   switch(node->getDataType().getVectorElementType())
+     {
+     case TR::Int8:
+       return TR::TreeEvaluator::inlineVectorBinaryOp(node, cg, TR::InstOpCode::vmaxsb);
+     case TR::Int16:
+       return TR::TreeEvaluator::inlineVectorBinaryOp(node, cg, TR::InstOpCode::vmaxsh);
+     case TR::Int32:
+       return TR::TreeEvaluator::inlineVectorBinaryOp(node, cg, TR::InstOpCode::vmaxsw);
+     case TR::Int64:
+       return TR::TreeEvaluator::inlineVectorBinaryOp(node, cg, TR::InstOpCode::vmaxsd);
+     case TR::Float:
+       return TR::TreeEvaluator::inlineVectorBinaryOp(node, cg, TR::InstOpCode::vmaxfp);
+     case TR::Double:
+       return TR::TreeEvaluator::vmaxDoubleHelper(node, cg);
```

Comment:
The reason the double helper is needed is because the VSX instruction for vmax for floating point vectors (`xvmaxsp` and `xvmaxdp`) don't match the behavior of the java library's Vector.max(). Specifically, when one of the operands is NaN, they return different results.

For FloatVectors, the VMX instruction, `vmaxfp` matches up with the java library version, so we can just use that single instruction instead. However, there isn't a VMX vector max instruction for DoubleVectors, so a multi-instruction sequence is needed to account for the NaN case.

Of course, all of this applies to vmin as well.

Reply:
OK, sorry, I thought it would be odd if those instructions had different semantics between float/double types.